sympy/combinatorics/homomorphisms.py

```diff
@@ -0,0 +1,225 @@
+        '''
+        Check if the homomorphism is injective
+
+        '''
+        return self.kernel().order() == 1
+
+    def is_surjective(self):
+        '''
+        Check if the homomorphism is surjective
+
+        '''
+        from sympy import S
+        im = self.image().order()
+        oth = self.codomain.order()
+        if im == S.Infinity and oth == S.Infinity:
+            return None
+        else:
+            return im == oth
+
+    def is_isomorphism(self):
+        '''
+        Check if `self` is an isomorphism.
+
+        '''
+        return self.is_injective() and self.is_surjective()
+
+    def is_trivial(self):
+        '''
+        Check is `self` is a trivial homomorphism, i.e. all elements
+        are mapped to the identity.
+
+        '''
+        return self.image().order() == 1
+
+def homomorphism(domain, codomain, gens, images=[]):
+    '''
+    Create (if possible) a group homomorphism from the group `domain`
+    to the group `codomain` defined by the images of the domain's
+    generators `gens`. `gens` and `images` can be either lists or tuples
+    of equal sizes. If `gens` is a proper subset of the group's generators,
+    the unspecified generators will be mapped to the identity. If the
+    images are not specified, a trivial homomorphism will be created.
+
+    If the given images of the generators do not define a homomorphism,
+    an exception is raised.
+
+    '''
+    if isinstance(domain, PermutationGroup):
+        raise NotImplementedError("Homomorphisms from permutation groups are not currently implemented")
+    elif not isinstance(domain, (FpGroup, FreeGroup)):
+        raise TypeError("The domain must be a group")
+    if not isinstance(codomain, (PermutationGroup, FpGroup, FreeGroup)):
+        raise TypeError("The codomain must be a group")
+
+    generators = domain.generators
+    if any([g not in generators for g in gens]):
+        raise ValueError("The supplied generators must be a subset of the domain's generators")
+    if any([g not in codomain for g in images]):
+        raise ValueError("The images must be elements of the codomain")
+
+    if images and len(images) != len(gens):
+        raise ValueError("The number of images must be equal to the number of generators")
+
+    gens = list(gens)
+    images = list(images)
+    images.extend([codomain.identity]*(len(generators)-len(images)))
+    gens.extend([g for g in generators if g not in gens])
+    images = dict(zip(gens,images))
+
+    if not _check_homomorphism(domain, images, codomain.identity):
+        raise ValueError("The given images do not define a homomorphism")
+    return GroupHomomorphism(domain, codomain, images)
+
+def _check_homomorphism(domain, images, identity):
+    rels = domain.relators
+    def _image(r):
+        if r.is_identity:
+            return identity
+        else:
+            w = identity
+            i = 0
+            while i < len(r):
+                power = r.array_form[i][1]
```

Comment:
Both forms of `r` are used alternately in the loop, array form on this line and letter form for `r[i]` on the next line. It might be possible to avoid the repeated transformations by using a loop like this: `for gen, pow in r:` with `r` in array form.